internal/controller/memcached_controller.go

```diff
@@ -75,12 +76,106 @@ func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+	version := memcached.Spec.Version
+	currentVersion := memcached.Status.CurrentVersion
+	checkedVersion := memcached.Status.CheckedVersion
+
+	if currentVersion == "" && version != checkedVersion {
+		// Initial case: no version is deployed yet.
+		// Check if version in the spec is a valid version, and write it to the status.
+		valid, err := isValidVersion(version)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Unexpected error checking version")
+			return ctrl.Result{}, err
+		}
+
+		memcached.Status.CheckedVersion = version
+
+		if valid {
+			// Version is valid, use it as the initial version.
+			memcached.Status.CurrentVersion = version
+		} else {
+			log.Info("5a. Update the memcached version. Initial version is not valid", "memcached.Spec.Version", memcached.Spec.Version)
+		}
+
+		if memcached.Status.Nodes == nil {
+			memcached.Status.Nodes = []string{}
+		}
+
+		err = r.Status().Update(ctx, memcached)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Failed to update status")
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if version != currentVersion && version != checkedVersion {
+		// Second case: a new version was applied.
+		// Check if version in the spec is a valid version, then check if the version change is a minor upgrade, if yes write the new version to the status.
+		valid, err := isValidVersion(version)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Unexpected error checking version")
+			return ctrl.Result{}, err
+		}
+
+		isMinorUpgrade, err := checkMinorUpgrade(currentVersion, version)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Unexpected error comparing versions")
+			return ctrl.Result{}, err
+		}
+
+		if !valid || !isMinorUpgrade {
+			// New version is not valid or not a minor upgrade, write to the status that it was already checked.
+			log.Info("5a. Update the memcached version. New version is not a valid minor upgrade", "memcached.Spec.Version", version, "memcached.Status.Version", currentVersion)
+
+			memcached.Status.CheckedVersion = version
+
+			if memcached.Status.Nodes == nil {
+				memcached.Status.Nodes = []string{}
+			}
+
+			err = r.Status().Update(ctx, memcached)
+			if err != nil {
+				log.Error(err, "5a. Update the memcached version. Failed to update status")
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// New version is valid and is a minor upgrade. Write it as the new version.
+		log.Info("5a. Update the memcached version. New version is a valid minor upgrade, updating status", "memcached.Spec.Version", version, "memcached.Status.Version", currentVersion)
+		memcached.Status.CheckedVersion = version
+		memcached.Status.CurrentVersion = version
+
+		if memcached.Status.Nodes == nil {
+			memcached.Status.Nodes = []string{}
+		}
+
+		err = r.Status().Update(ctx, memcached)
```

Comment:
Question: have you considered implementing this directly in the image upgrade logic in line 197? That way the indirection of writing to the status can be avoided

Reply:
I have an issue with this approach, which depends on some assumptions I made or features I strived for which are not necessarily required:
What if the user initially inputs an invalid version? Is it ok to apply a deployment with an invalid image, leading to errored pods? In that case, there's no need to discuss further.
If it is not desired, and there should not be a deployment if the user's initial version is invalid, then, the way I see it, in the loop the controller wouldn't be able to differentiate between the case of the deployment missing because of an invalid version or simply because it was never applied before, or because it was deleted by external factors.
The controller would have to always check the version first (which means HTTP request) on each reconcile in the case of a missing deployment, which is redundant work.
If I hadn't implemented the version check to do a HTTP request, it wouldn't be a big deal to do the work every time, but it does make it evident that it is not 100% "clean" that way.

Reply:
After considering it again, I think the logic with the checkedVersion field falls apart no matter what when considering that the deployment's image field can change externally as you mentioned in your other comment. I decided to remove it entirely. Premature optimization led me astray!